### ISSUE TITLE
`nextToken()` decodes with `llama_batch_get_one`: wrong positions and a KV-cell leak on iSWA models after prefix reuse past the sliding window

### ISSUE THREAD

**Issue by juliusbangert** [mybigday/llama.rn#396]
I've spent a few days chasing a collapse in our app (Gemma 4 E2B on iOS) where, part-way through a long chat, every reply came back with zero tokens and no stop reason, and it stayed that way until we cleared the cache. I got to the bottom of it and it's two related things in `cpp/rn-completion.cpp`, both specific to sliding-window (iSWA) models with prefix reuse. I have a fix for both that we've been running on device, and I'm happy to turn it into a PR.

### Environment

- llama.rn 0.12.9 (llama.cpp b10256) and 0.13.0-rc.4 (b11010) — `nextToken()` and `loadPrompt` are the same in both.
- Gemma 4 E2B IT (Unsloth QAT `UD-Q4_K_XL`): `sliding_window = 512`, 15 KV layers of which 12 are sliding and 3 global.
- `n_ctx` 6144, `n_batch` 512, default `swa_full` (false, so the sliding cache is `n_swa + n_ubatch` = 1,024 cells), q4_0 KV, one sequence. iOS/Metal, but nothing here is backend-specific.

### Symptom

In a test that carries the cache across ~27 turns of a growing conversation, replies start coming back empty: `tokens_predicted: 0`, and `stopped_eos`, `stopped_limit`, `stopped_word`, `truncated`, `context_full`, `interrupted` all false. The native log says:

```
decode: failed to find a memory slot for batch of size 402
decode: failed to find a memory slot for batch of size 512
```

Once it starts it's sticky — every later completion fails until `llama_memory_clear`. That's with prompts of 1–2.3k tokens in a 6,144-cell cache, so it isn't context overflow.

### What's happening

`nextToken()` evaluates the prompt tail with

```cpp
llama_decode(parent_ctx->ctx, llama_batch_get_one(&embd[n_past], n_eval))
```

`llama_batch_get_one` gives the batch no positions, so `llama_batch_allocr::init` derives them from `memory->seq_pos_max(seq) + 1`. For an iSWA memory that is the **sliding** cache's maximum (`llama-kv-cache-iswa.cpp`):

```cpp
llama_pos llama_kv_cache_iswa::seq_pos_max(llama_seq_id seq_id) const {
    return kv_swa->seq_pos_max(seq_id);
}
```

The sliding cache only keeps the last ~1,024 positions; older ones are purged as generation goes on. So:

1. Turn A ends at position ~1,300. The sliding cache has purged everything below ~800.
2. Turn B shares the first ~400 tokens with A. `loadPrompt` sets `n_past = 400` and calls `seq_rm(0, 400, -1)`. The sliding cache now holds nothing for seq 0, so `seq_pos_max` is −1.
3. The tail is decoded at positions **0, 1, 2, …** instead of 400, 401, …. The base cache already has cells at 0–399, so each of those positions gets a second cell. They sit below `n_past`, so no later `seq_rm(0, n_past, -1)` reaches them; they pile up until both caches are full. The turn itself is also decoded at the wrong positions.
4. `embd`/`n_past` on the llama.rn side carry on counting from 400, so the bookkeeping and the cache disagree from then on.

Whenever the divergence is inside the window the derived positions happen to be right, which is why short chats and per-turn-cleared tests never show it. The `probeStateCache()` comment already describes the gap:

> TODO: SWA reuse past a slid window (deep edit in a chat longer than the window) is unguarded here; a fix would need a pos_min trigger, not seq_rm.

This is what that gap does in practice.

How I measured it: `saveSession` → `llama_state_save_file` writes every non-empty cell of both caches, so the file size is the occupancy. With this model's geometry a base cell is ~1.7 KB and a sliding cell ~3.5 KB, and on clean turns the file size matched `tokens × bytes-per-cell` to within a few cells. On the turns that diverged far back it grew by +400…+1,100 base cells each — +277 → +699 → +1,099 → +1,542 → +1,954 → +2,808 → +3,559 cells above the live sequence, then both caches full, then the failure.

### Fix, part 1 — give the batch its positions

`n_past` is llama.rn's own record of what the cache holds and has just been used for `seq_rm`, so it's the right source. The MTP path in the same file already builds batches this way.

```cpp
-        if (llama_decode(parent_ctx->ctx, llama_batch_get_one(&embd[n_past], n_eval)))
+        llama_batch batch = llama_batch_init(n_eval, 0, 1);
+        for (int i = 0; i < n_eval; ++i) {
+            common_batch_add(batch, embd[n_past + i], n_past + i, { 0 }, i == n_eval - 1);
+        }
+        const int decode_ret = llama_decode(parent_ctx->ctx, batch);
+        llama_batch_free(batch);
+        if (decode_ret)
```

Behaviour is identical whenever the derived positions were right (non-SWA models, and SWA within the window): the explicit positions equal the derived ones and only the last token requests logits, as before. It's one `llama_batch_init`/`free` per decode call; a pre-allocated batch on the completion object would remove even that. The `llama_encode` call for encoder models a few hundred lines up uses the same helper and should probably get the same treatment for consistency (not exercised by us).

### Fix, part 2 — the same reuse leaves the sliding layers blind

With positions fixed, the leak was gone (occupancy flat to within ±5 cells over three runs) and a second problem showed up in its place. After a long previous sequence, `seq_rm(n_past, -1)` leaves the sliding cache with **no cells for the reused prefix** — they were purged as the sequence grew. The 12 sliding layers then attend over an empty window for the whole turn: no system prompt, no tool declarations, no history. Only the 3 global layers see them. On device that turned into the model narrating tool calls instead of making them — "Fetching current weather data for London." — on every tool turn of a run whose previous sequence had passed ~1.4k tokens.

llama-server already guards this (the `pos_min_thold` check in `server-context.cpp`, which forces a full reprocess when `pos_min` is past `n_past − n_swa`, referencing ggml-org/llama.cpp#13194). `loadPrompt` doesn't. Minimal port, after the `seq_rm`:

```cpp
if (n_past > 0) {
    const int32_t n_swa = llama_model_n_swa(parent_ctx->model);
    if (n_swa > 0) {
        const llama_pos swa_min = llama_memory_seq_pos_min(kv, 0);   // iSWA: the sliding cache's min
        const llama_pos needed_from = std::max<llama_pos>(0, n_past - n_swa);
        if (swa_min < 0 || swa_min > needed_from) {
            LOG_INFO("prefix reuse past the sliding window (n_past=%d, swa_min=%d, n_swa=%d): full reprocess", n_past, swa_min, n_swa);
            llama_memory_clear(kv, false);
            clearStateCheckpoints();
            n_past = 0;
        }
    }
}
```

It only fires when the divergence is deeper than the window — never on append turns, never with `swa_full`, never on non-SWA models — and costs one prefill when it does. A partial reprocess from `n_past − n_swa` wouldn't be correct because the states at those positions depend on the window before them.

### Results on device

Same 27-case suite, three consecutive runs on one context, occupancy probed after every completion:

|                                          | before               | part 1 only | parts 1 + 2      |
| ---------------------------------------- | -------------------- | ----------- | ---------------- |
| tool calls made (of 12)                  | 11 / 9 / 4           | 9 / 9 / 4   | **12 / 12 / 11** |
| widgets rendered (of 27)                 | 19 / 15 / 7          | 10 / 13 / 6 | **21 / 22 / 22** |
| empty replies (0 tokens, no stop reason) | 18 of 27 in runs 2–3 | 0           | **0**            |
| occupancy vs live tokens                 | +3,500 cells, rising | flat        | flat, +5…+69     |

A full 14-suite batch (~700 turns) on the fixed build afterwards: three defects we'd been attributing to the model went away with it, and the timing suite's TTFT didn't move (2.67–2.72 s), so the reprocess isn't costing anything measurable at the divergence depths a chat reaches.

### Minimal reproduction

Any SWA model, default `swa_full`, one context:

1. Prompt A = a ~400-token system prompt + a user message; generate ≥ ~1,000 tokens.
2. Prompt B = the same system prompt + a different user message. Complete.
3. `saveSession` and compare the file size with `tokens_cached × bytes-per-cell` — it holds ~400 extra base cells.
4. Repeat with fresh user messages; each iteration adds cells; eventually the completion returns 0 tokens with no stop flag and the log says `failed to find a memory slot`.

I'll open a PR with both parts if you're happy with the approach. The diffs above are what we run as a patch, on 0.12.9 and now on 0.13.0-rc.4.
